pkgs/development/julia-modules/python/extract_artifacts.py

```diff
@@ -0,0 +1,99 @@
+
+import json
+from pathlib import Path
+import multiprocessing
+import subprocess
+import sys
+import toml
+import yaml
+
+import dag
+
+dependencies_path = Path(sys.argv[1])
+closure_yaml_path = Path(sys.argv[2])
+julia_path = Path(sys.argv[3])
+extract_artifacts_script = Path(sys.argv[4])
+extra_libs = json.loads(sys.argv[5])
+out_path = Path(sys.argv[6])
+
+with open(dependencies_path, "r") as f:
+  dependencies = yaml.safe_load(f)
+  dependency_uuids = dependencies.keys()
+
+with open(closure_yaml_path, "r") as f:
+  # Build up a map of UUID -> closure information
+  closure_yaml_list = yaml.safe_load(f) or []
+  closure_yaml = {}
+  for item in closure_yaml_list:
+    closure_yaml[item["uuid"]] = item
+
+  # Build up a dependency graph of UUIDs
+  closure_dependencies_dag = dag.DAG()
+  for uuid, contents in closure_yaml.items():
+    if contents.get("depends_on"):
+      closure_dependencies_dag.add_node(uuid, dependencies=contents["depends_on"].values())
+
+with open(out_path, "w") as f:
+  f.write("{ lib, fetchurl, glibc, pkgs, stdenv }:\n\n")
+  f.write("rec {\n")
+
+  def process_item(item):
+    uuid, src = item
+    lines = []
+    artifacts = toml.loads(subprocess.check_output([julia_path, extract_artifacts_script, uuid, src]).decode())
+    if not artifacts: return f'  uuid-{uuid} = {{}};\n'
+
+    lines.append(f'  uuid-{uuid} = {{')
+
+    for artifact_name, details in artifacts.items():
+      if len(details["download"]) == 0: continue
+      download = details["download"][0]
+      url = download["url"]
+      sha256 = download["sha256"]
+
+      git_tree_sha1 = details["git-tree-sha1"]
+
+      depends_on = set()
+      if closure_dependencies_dag.has_node(uuid):
+        depends_on = set(closure_dependencies_dag.get_dependencies(uuid)).intersection(dependency_uuids)
+
+      other_libs = extra_libs.get(uuid, [])
+
+      fixup = f"""fixupPhase = let
+          libs = lib.concatMap (lib.mapAttrsToList (k: v: v.path))
+                               [{" ".join(["uuid-" + x for x in depends_on])}];
+          in ''
+            find $out -type f -executable -exec \
+              patchelf --set-rpath \$ORIGIN:\$ORIGIN/../lib:${{lib.makeLibraryPath (["$out" glibc] ++ libs ++ (with pkgs; [{" ".join(other_libs)}]))}} {{}} \;
```

Comment:
Did you find it necessary for `$ORIGIN` and `$ORIGIN/../lib` have take precedence over nixpkgs' glibc and libs? This, in particular, means that any vendored libraries potentially distributed with the package would be loaded first, even when there's one we've built from source

Reply:
No, I didn't try a different precedence for the `$ORIGIN` paths. It seemed to me that if some library is vendored alongside the given artifact, there's probably a reason for it--the artifact expects to be linked against that version, and this is more likely to succeed than if we link against whatever version is in Nixpkgs. I agree that this feels strange if the artifact should choose to vendor something like `glibc`, but in principle I think even that is something we should allow Julia's artifacts system to do if it wants.

Reply:
> there's probably a reason for it-

They might just assume they won't be able to find it on the host...


Not a blocker, but I would suggest to consider giving nixpkgs the priority. This should be somewhat safe, assuming version sonames: e.g. if a package depends on `libcublas.so.11` and vendors a copy, and we happen to give it `libcublas.so.12`, the linker would just skip our version and fall back to one from `$ORIGIN`. On the other hand, using packages from nixpkgs has the mentioned benefit that we rely on prebuilt stuff less and that we reuse patches applied elsewhere in the package set...

Reply:
> we reuse patches applied elsewhere in the package set

Potentially at the expense of losing any patches the Julia ecosystem has applied. But yeah, I agree this is worth trying. It might even be the key to getting CUDA to work :)